language-plutus-core/src/Language/PlutusCore/Subst.hs

```diff
@@ -95,20 +97,49 @@ typeSubstTyNames
     -> Type tyname uni ann
 typeSubstTyNames = purely typeSubstTyNamesM
 
--- | Naively substitute names using the given functions (i.e. do not substitute binders).
+-- | Naively substitute names using the given function (i.e. do not substitute binders).
 termSubstNames
     :: (name ann -> Maybe (Term tyname name uni ann))
     -> Term tyname name uni ann
     -> Term tyname name uni ann
 termSubstNames = purely termSubstNamesM
 
--- | Naively substitute type names using the given functions (i.e. do not substitute binders).
+-- | Naively substitute type names using the given function (i.e. do not substitute binders).
 termSubstTyNames
     :: (tyname ann -> Maybe (Type tyname uni ann))
     -> Term tyname name uni ann
     -> Term tyname name uni ann
 termSubstTyNames = purely termSubstTyNamesM
 
+-- | Monadically substitute *free* names using the given function.
+termSubstFreeNamesA
```

Comment:
I find it a little weird that this one is an `A` rather than an `M` like the others, but I guess if you can get away with the lesser constraint you might as well!

Reply:
It's not just about the lesser constraint, it also immediately gives the reader an idea of how the algorithm works. Namely that we do not recurse after calling `f` over a name, which is an option given that the result of this operation is a `Term`. And it's clear that we don't do that, because we'd have to be in a monad for that.

In general, I find it helpful to distinguish between applicatively and monadically effectful functions (I'll also update the comment to say "Applicatively"), especially given that certain functions have both the variants (e.g. `foldMapA` vs `foldMapM` -- the former is good for a lazy applicative and a non-constant-space monoid, the later is good for a strict monad and a constant-space monoid).

Reply:
I guess I wonder whether the other ones could be `Applicative`.

Reply:
There's `substTyVarA` and `substVarA`. Others can't be applicatively effectful, because `transformM` that they use under the hood is monadic as it has a similar behaviour to

> recurse after calling `f` over a name, which is an option given that the result of this operation is a Term